(maint) Use a module loader within the Puppetfile class

justinstoller opened the pull request:
This series of commits moves the logic for loading the module content from a Puppetfile and into a new concept called a ModuleLoader. The idea being to standardize the interface for loading modules, since there are multiple ways to load module specifications (eg, puppetfile, exec, yamldir).

This approach would allow Environments to no longer care _how_ modules were loaded but simply call `load` on any ModuleLoaders they are configured to use. The environment would then have a canonical list of all its environments (instead of deferring to the Puppetfile class when available) and be able to pass all of them to ContentSynchronizer at once (currently only Puppetfile content can be concurrent synced). As well as a list of managed directories, desired content, and purge exclusions for usage with the Cleaner util.

Refactoring the environment to use Module Loaders is a somewhat larger refactor and may break internal APIs. This changeset only refactors the Puppetfile to use the ModuleLoader under the hood. This should be a backwards compatible refactor save that `load` now returns the hash of loaded content information rather than `true` (we only check for it's truthiness - not a hard type check, so it remains backwards compatible for our uses). 


Please add all notable changes to the "Unreleased" section of the CHANGELOG in the format:
```
- (JIRA ticket) Summary of changes. [Issue or PR #](link to issue or PR)
```

justinstoller committed:
(maint) Move Puppetfile::DSL to new location

justinstoller committed:
(maint) Move module loading from Puppetfile to module loader

justinstoller committed:
(maint) Move puppetfile purge info to module loader

This moves the information gathering required to purge at the Puppetfile
level (which is a collection of module info) into the
ModuleLoader::Puppetfile. The existing Puppetfile logic defers to the
new puppetfile module loader. Several tests are commented out because
stubbing the information from the module loader is difficult from the
puppetfile test. These tests are commented out to provide a guide for
the next commit, which will flesh out testing of the puppetfile module
loader.

justinstoller committed:
(maint) Add tests for Puppetfile module loader

This commit audits the R10K::Puppetfile tests and moves/rewrites those
that are applicable to the R10K::ModuleLoader::Puppetfile tests. It also
makes some refactors to both for easier testing.

One notable behavior change in R10K::Puppetfile is that the `load`
method will now return the loaded content (which is truthy) rather than
the boolean true.

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -4,6 +4,11 @@ CHANGELOG
+3.9.1
+-----
+
+- Invalid module specifications in a Puppetfile will cause the R10K run to abort earlier than before. Prior to this release the R10K run would complete, sync all other modules, and return an exit code of one, R10K will now stop syncing modules and abort immediately.
```
Is there a ticket associated with this change? If so, we should include it. If not, we should at least link to this PR for context.

mwaggett replied:
```suggestion
- Invalid module specifications in a Puppetfile will cause the R10K run to abort earlier than before. Prior to this release, the R10K run would complete, sync all other modules, and return an exit code of 1. R10K will now stop syncing modules and abort immediately.
```

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -4,6 +4,11 @@ CHANGELOG
 Unreleased
 ----------
 
+3.9.1
```
Should this be a feature release? I could be convinced either way.

justinstoller replied:
No, this is an internal refactor.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+        :managed_directories, :desired_contents, :purge_exclusions
+
+      # @param basedir [String] The path that contains the moduledir &
+      #     Puppetfile by default. May be an environment, project, or
```
How is the `basedir` used if the `moduledir` and Puppetfile are elsewhere?

justinstoller replied:
Prior to this you could pass relative paths for these, however I took that out based on how convoluted the original implementation was. I've added that ability back (which leverages the basedir more thoroughly) and cleaned up the implementation.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+          managed_directories: @managed_directories,
+          desired_contents: @desired_contents,
+          purge_exclusions: @purge_exclusions
+        }.freeze
```
Is this `freeze` necessary if we're not saving this hash as a variable anywhere?

justinstoller replied:
This isn't necessary. I was toying with different refactors that would make this class a one-shot loader, or would return an immutable object that contained loaded content. I made the results of loading immutable as a start and without a strong push towards one implementation or the other I haven't change it yet.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+      ##
+
+      # @param [String] forge
+      def set_forge(forge)
```
Any reason for this not to just be an `attr_accessor`?

justinstoller replied:
I think it's good especially call out that this, `set_moduledir` and `add_module` are part of an interface contract with the DSL (vs, making an `attr_writer` and updating the DSL to call that).

mwaggett replied:
hmm okay. so then could you maybe update your comment to say that? as it is, I don't really see why we couldn't just update the DSL to call an `attr_writer`.

justinstoller replied:
What - in addition to the comment four lines above this - would like to see?

mwaggett replied:
Is there a reason we're treating the DSL's usage as an API contract? i.e. _why_ can't we update the DSL to call a different method?

Or, to put it another way, why is it especially important to note that these methods are used by the DSL?

Magisus replied:
I'm with Justin, I think the clarity of having them be "real" methods helps with realizing that they're some kind of API. There's too much wishy-washiness as it is with that. In addition, not all of them are pure setters. So it's kind of nice to see them as a set here.

Plus, I feel like updating the usage of the DSL class is outside the scope of this, besides.

mwaggett replied:
yeah, that's fine, I just want it to be clear that we're treating this as an API contract. so the comment could say something like
> The `set_forge`, `set_moduledir`, and `add_module` methods must be defined as part of an API contract with the DSL class.`

mwaggett replied:
Am I making sense?

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+      #   Hashes of the shape
+      #     { version: <String or Symbol> }
+      #
+      def add_module(name, args)
```
Can we call this param something more descriptive than `args`? The yardoc also doesn't really tell me what that param is for.

justinstoller replied:
Sure!

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+
+      # @param [String] name
+      # @param [Hash, String, Symbol] args Calling with anything but a Hash is
+      #   deprecated. The DSL will now convert String and Symbol versions to
```
Is this a user-facing deprecation?

justinstoller replied:
This is a deprecation that would affect those writing their own Puppetfile parsers but want to use most of our internals.

Previously, we would pass whatever crap users put after `mod` in a Puppetfile down to the Module implementations and hope they could figure out what was what. This will, in a future version, require Puppetfile DSL implementations to convert all that nonsense into a consumable hash with some consistent structure.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+        # Do not load modules if they would conflict with the attached
+        # environment
+        if @environment && @environment.module_conflicts?(mod)
+          mod = nil
```
I don't think this is necessary?

justinstoller replied:
Probably not, I think we used to unconditionally add the module and later on compact `@modules`. I'll remove it.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+      def validate_no_duplicate_names!(modules)
+        dupes = modules
+                .group_by { |mod| mod.name }
+                .select { |_, v| v.size > 1 }
```
Can we use something more descriptive than `v`? maybe `modules_by_name` or something, so it's clear what's being counted?

justinstoller replied:
sure

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+                .select { |_, v| v.size > 1 }
+                .map(&:first)
+        unless dupes.empty?
+          msg = _('Puppetfiles cannot contain duplicate module names.')
```
Can you remind me what wrapping this in `_()` does? It's not very easy to google haha

Magisus replied:
This is for i18n, it's the string extraction function, like the `trs`/`tru` from our Clojure.

mwaggett replied:
ahh gotcha, no wonder I didn't find it by googling 😅

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+     private
+
+      # @param [Array<R10K::Module>] modules
+      def validate_no_duplicate_names!(modules)
```
Why does this have a `!`? It doesn't modify `modules` at all, does it?

Magisus replied:
This is a clojurism that doesn't mean the same thing in Ruby, agree it should be dropped.

justinstoller replied:
A method that ends in a bang means that the method is more dangerous or surprising than one would expect. Typically in core classes it is used for methods that mutate the receiver. It is also used for methods that raise on failure instead of return data, or methods that may not be chainable in ways that similar methods are. Bang methods are always paired with a non-bang version. https://stackoverflow.com/questions/612189/why-are-exclamation-marks-used-in-ruby-methods

In Clojure the bang means a method should not be used in a transaction - I think we use it especially in file-sync for things that modify the file system (though I wouldn't be surprised if we use it in functions that raise as well). https://stackoverflow.com/questions/20606249/when-to-use-exclamation-mark-in-clojure-or-lisp

The exclamation mark is there because its a validation function that raises rather than returns data, which I think can be an accurate usage in Ruby, but it isn't paired with a non-bang version, so I agree it should be dropped.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+          pn = Pathname.new(File.join(@basedir, path))
+        end
+
+        # .cleanpath is as good as we can do without touching the filesystem.
```
as good as we can do for what? (I looked it up, so I know what `.cleanpath` does now, but since we're leaving this comment, I think it would be good to specify.)

justinstoller replied:
updated.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+        pn.cleanpath.to_s
+      end
+
+      def validate_install_path!(path, modname)
```
Same question about `!` in the method name - it doesn't look like this modifies any of the args?

justinstoller replied:
Ditto to above. Raising without returning data is a valid reason to append an exclamation mark to a method name, but only if there's also a "safe" variant.

In that spirit I've also removed the exclamation mark from `load!` (the `R10K::Puppetfile` object has both `load!` and `load` but this object does not).

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+
+      def real_basedir
+        Pathname.new(@basedir).cleanpath.to_s
+      end
```
Is there a reason to have this as a separate method as opposed to just setting `@basedir = Pathname.new(basedir).cleanpath.to_s` on initialization?

justinstoller replied:
I didn't like this before and went ahead and refactored it in response to your comment regarding basedir/moduledir/puppetfile. Let me know what you think.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+      end
+
+      # For testing purposes only
+      def puppetfile_content(path)
```
I still don't really see the purpose of this.. Why not just call `File.read` directly if this is just for testing?

justinstoller replied:
So specs can do:
```rb
test_puppetfile_content = <<EOPF
moduledir "my/moduledir"

mod "my-module", "1.0"
EOPF

expect(puppetfile).to receive(:puppetfile_content).and_return(test_puppetfile_content)
puppetfile.load
```

mwaggett replied:
yeah, I'm just not really understanding why that's even necessary. This isn't acting on anything within this class.. 

In this example, you're acting on `puppetfile`, but this is the `loader` class.. Should this method go in the `R10K::Puppetfile` class? Or should this not take a param and instead call `File.read(@puppetfile)`?

justinstoller replied:
So the fact that it takes a param instead of acting on an instance variable is the confusing part? It's six of one, half dozen of the other to me. Though I typically like methods parameterized as much as possible, which is why it has a param.

Here is an example of how this is used in a test: https://github.com/puppetlabs/r10k/pull/1167/files#diff-1bc365ddca4e6e4a7ba95ff4d655c036faa16ec1025e35bc559476265a5719aaR292

mwaggett replied:
No, that's not what bothering/confusing me.

I was mostly objecting to the premise of adding a class method just for testing purposes, especially since I wasn't really understanding how calling it in the tests was actually doing anything, but now that I look again, I see that this method actually _is_ being used within the `load` method, not just in tests, in which case I think you should remove this comment, since it clearly confused me a lot lol.

Magisus replied:
I agree this makes things better for testing, it gets messy stubbing things like `File.read` because once you do it once in a test, you have to do it for all instances of it being called, which often means stubbing a lot more than is strictly necessary. So far as I can tell, it _is_ called up above, where it's being passed the `puppetfile` instance var. The little abstraction over `File.read` in that context is kinda nice actually, IMO.

mwaggett replied:
yeah, I'm totally fine with this method now that I see how it's being used. The "for testing purposes only" comment totally threw me off.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+
+      # Returns an array of the full paths to all the content being managed.
+      # @return [Array<String>]
+      def determine_desired_contents(managed_content)
```
possibly out of scope for this if you just copied it from elsewhere, but I don't find `desired_contents` to be a particularly informative name for what we're actually collecting here.. `managed_content` and `managed_directories` are not much better..

justinstoller replied:
Yeah, I don't either, but `managed_directories`, `desired_contents`, and `purge_exclusions` make up an interface for interacting with the Purgeable mixin that's used in several places. I think `determine_<interface name>` (unless you have a better word for "determine") is good.

mwaggett replied:
that works for me!

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+        end
+      end
+
+      def determine_purge_exclusions(managed_dirs)
+        if environment && environment.respond_to?(:desired_contents)
+          managed_dirs + environment.desired_contents
+        else
+          managed_dirs
+        end
+      end
```
Does this mean we exclude all modules from purging??

justinstoller replied:
Yeah, a Puppetfile "manages" the moduledir and will purge anything inside it that _isn't_ a "module".

mwaggett replied:
okay that makes sense

Magisus reviewed (approved):
So just to sum up, the goal of this is to establish the new API of a ModuleLoader, but deliberately does not update any consumers of module loading to use it; everything is still delegated via the `Puppetfile` class.

I think this makes sense as an intermediate step. Do you anticipate being able to switch over to calling things via this new API directly, at least internally, without breaking things for users? Is Bolt the only known consumer of r10k as a library that we need to worry about breaking?

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +72,109 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
+    @environment     = options.delete(:environment)
 
     @overrides       = options.delete(:overrides) || {}
+    @default_branch_override = @overrides.dig(:environments, :default_branch_overrides)
```
Should this be `:default_branch_override` singular? If so, seems not great if tests didn't catch that..

justinstoller replied:
Yeah it should be. Though what this is doing is saving the initialized value for the conditional that warns below. With this bug we will always warn. I've corrected this and added tests around the default_branch_override behavior.

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +72,109 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
-    return true if self.loaded?
-    if File.readable? @puppetfile_path
-      self.load!(default_branch_override)
+  def load(default_branch_override = NotGiven)
```
Why set up this `NotGiven` object instead of continuing to use `nil`?

justinstoller replied:
I was unclear if `nil` had meaningful semantics as a `default_branch_override`. I don't think so, but I wanted to validate the case between a `nil` passed explicitly into `load` and the default value for it. I've removed it.

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +72,109 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
-    @default_branch_override = default_branch_override
+  def load!(dbo = NotGiven)
+    if (dbo != NotGiven) && (dbo != @default_branch_override)
+      logger.warn("Mismatch between passed and initialized default branch overrides, preferring passed value.")
```
I think this message could be a little more user-friendly. When would this situation arise? Where would these two values be coming from?

justinstoller replied:
This is largely for debugging/development purposes and I don't expect users to come across it in the wild.

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +72,109 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
 
-  def load!(default_branch_override = nil)
-    @default_branch_override = default_branch_override
+  def load!(dbo = NotGiven)
```
I'd love if we left `default_branch_override` instead of `dbo`.

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +72,109 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
+  def load!(dbo = NotGiven)
+    if (dbo != NotGiven) && (dbo != @default_branch_override)
+      logger.warn("Mismatch between passed and initialized default branch overrides, preferring passed value.")
+      @loader.default_branch_override = dbo
```
Do we still want to set this if we can't read the Puppetfile and therefore never actually load?

justinstoller replied:
I want to set this at the ModuleLoader initialization and have passing this into deprecated. The warning above is largely for debugging/API usage. Likewise, the `NotGiven` construct was to ensure `nil` didn't have special semantics as an override. I've added docs deprecating this signature.

mwaggett commented on spec/unit/action/deploy/module_spec.rb:
```diff
@@ -202,14 +202,20 @@
         end
 
-        pf.modules.each do |mod|
+        loader.modules.each do |mod|
```
I'm confused about why this needed to change.. Shouldn't we still be able to act on the Puppetfile the way we were before? 
Are we intending to eventually get rid of the `Puppetfile` class altogether at some point? Even then, I feel like this change should go with that removal? Continuing to go through `Puppetfile` for now seems like it could be a good test of backwards compatibility?

Or am I just totally not understanding something?

justinstoller replied:
No, that makes sense. Though this (and the next test) aren't really testing Puppetfile compatibility, they really just have a bunch of setup so that we can test that when you call `r10k deploy module` with a list of modules and/or environments to deploy only those modules/environment combinations are deployed.

We just stub out all the Puppetfile interactions so we get realistic modules and can put expectations on their internal Repos.

Turns out the change itself wasn't needed and I changed it to `puppetfile`. I've also added comments explaining more.

mwaggett commented on spec/unit/action/deploy/module_spec.rb:
```diff
@@ -182,7 +182,7 @@
 
     let(:cache) { instance_double("R10K::Git::Cache", 'sanitized_dirname' => 'foo', 'cached?' => true, 'sync' => true) }
-    let(:repo) { instance_double("R10K::Git::StatefulRepository", cache: cache, resolve: 'main') }
+    let(:repo) { instance_double("R10K::Git::StatefulRepository", cache: cache, resolve: 'main', tracked_paths: []) }
```
What is `tracked_paths` for?

justinstoller replied:
Determining `desired_contents` on the environment which is now preemptively done in the `load` method rather than on demand during purging.

mwaggett replied:
I'm just not seeing how/where it's being used..

justinstoller replied:
When we call `determine_purge_exclusions` we ask the environment for its `desired_contents`. A Git backed environment will call `tracked_paths` on it's backing repo (here: https://github.com/puppetlabs/r10k/blob/main/lib/r10k/environment/git.rb#L90)

mwaggett replied:
ah, so previously we were getting away with not including `tracked_paths` here because this test used to not reach a point where we needed the `desired_contents`, yes? this is just some stubbing weirdness and a real `R10K::Git::StatefulRepository` would have always had `tracked_paths`, even before this PR?

Magisus replied:
Yeah.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+      subject do
+        R10K::ModuleLoader::Puppetfile.new(basedir: '/test/basedir/env',
+                                           moduledir: '/test/basedir/env/dist-modules',
+                                           puppetfile: '/test/basedir/env/puppetfile.prod',
```
Would it be worth using values that are not inside the `basedir`, just to really be sure that we're not twiddling things we shouldn't?

mwaggett replied:
Or are these required to be inside the `basedir`? I saw that we validate that a module's `install_path` is inside the `basedir`, but I didn't see any of that validation for the `moduledir` or `puppetfile`.

justinstoller replied:
I've refactored this test as part of addressing your first comment regarding basedir/moduledir/puppetfile parameters.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+        subject.add_module('puppet/test_module', '< 1.2.0')
+      }.to raise_error(RuntimeError, /module puppet\/test_module.*doesn't have an implementation/i)
+
+      expect(subject.modules.collect(&:name)).not_to include('test_module')
```
since we're using the same `subject` for all these tests, I'm confused about how the 'test_module' from the previous test wouldn't still be in `subject.modules` for this test..

justinstoller replied:
`subject` is just a distinguished name in the specs. It is effectively
```ruby
let(:subject) <block>
```
where `<block>` if not set would default to:
```rb
R10K::ModuleLoader::Puppetfile.new
```
since that is the class passed into the top level `describe` block.
(and `let`s are memoized within a nested test block but are reset between them)

mwaggett replied:
oh, so even though it's defined within the `describe`, each `it` block resets the value?

justinstoller replied:
yep!

mwaggett replied:
cool, thanks for explaining!

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+        git: 'git@example.com:puppet/test_module.git',
+      }
+
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', File.join(basedir, 'vendor'), module_opts, anything).and_call_original
```
I think we can remove this `allow`, right? since `validate_install_path!` gets called before we initialize the module, we would NOT expect this to get called at all, right?

justinstoller replied:
good catch!

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+        git: 'git@example.com:puppet/test_module.git',
+      }
+
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', File.join(basedir, 'vendor'), module_opts, anything).and_call_original
```
same re: removing this `allow`

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,190 @@
+
+        # Do not load modules if they would conflict with the attached
+        # environment
+        if @environment && @environment.module_conflicts?(mod)
```
Should we be warning at all in this case?

justinstoller replied:
I'm not clear enough on this behavior to know if users frequently ignore a large subset of Puppetfile content because of this (and are alright with that). The person who set users up with this workflow implemented it so I'm going to assume it's working okay for their use case. (Reid)

mwaggett replied:
fair, maybe at least some debug logging would be useful?

justinstoller replied:
Looking into it more it seems that the `module_conflicts?` method handles that already https://github.com/puppetlabs/r10k/blob/main/lib/r10k/environment/with_modules.rb#L52

mwaggett replied:
ahh nice, thanks for finding that.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+    it 'should disable and not add modules that conflict with the environment' do
+      env = instance_double('R10K::Environment::Base')
+      mod = instance_double('R10K::Module::Base', name: 'conflict', origin: :puppetfile)
+      allow(mod).to receive(:origin=).and_return(nil)
```
What is this for?

justinstoller replied:
It was a stub, I updated the `instance_double` line above to set it with the reader.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+      allow(env).to receive(:'module_conflicts?').with(mod).and_return(true)
+
+      allow(R10K::Module).to receive(:new).with('test', anything, anything, anything).and_return(mod)
+      expect { loader.add_module('test', {}) }.not_to change { loader.modules }
```
It's confusing to me that we're using 'test' as the name here, but we use 'conflict' inside `mod`.

justinstoller replied:
good catch, I updated it.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+    subject { R10K::ModuleLoader::Puppetfile.new(basedir: '/test/basedir') }
+
+    it 'includes managed_directories' do
+      expect(subject.send(:determine_purge_exclusions, managed_dirs)).to match_array(managed_dirs)
```
Is the reason you need to use `send` because `determine_purge_exclusions` is private? 

I feel like it might make more sense to test the behavior we want to see rather than reaching into private methods? Like, as is, this test doesn't seem very valuable to me..

justinstoller replied:
That's fair. These were existing tests on Puppetfile that I updated minimally to work with the ModuleLoader, I'll re-work them to be a bit higher level if I can.

justinstoller replied:
Did I not do this?

justinstoller replied:
Looks like no. I believe after I push my local changes up, this will be the one outstanding change that I still need to make, however I won't get to it till next week.

mwaggett replied:
Is this still something you're planning to do as part of this PR?

Magisus replied:
Fwiw, as much as I agree that this test isn't super useful, I also think it would be good to get this in, so if this is the only pending comment, I'm inclined to leave writing new tests for another time. We'll get better validation running it in real environments.

mwaggett replied:
Do we want/need a ticket for writing better automated tests?

The only other pending comments were around code comments that I found confusing, one of them being straight up not true (the one about the `puppetfile_content` method). so, not the biggest deal, but could be extra confusing for future us. ¯\_(ツ)_/¯

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+      allow(subject).to receive(:puppetfile_content).and_return('')
+    end
+
+    it 'returns an array of paths that can be purged' do
```
Should this reference purging?? I was gonna blame copy pasta, but it doesn't even look like this is used in the `purge_exclusions` test.. ¯\_(ツ)_/¯

justinstoller replied:
I think it makes sense to reference purging here though wording is weird and the test does a bit more than it really needs to. This is part of the inputs to the Purgeable mixin, which will find everything in `managed_directories` (here the moduledir) and then purge everything except `desired_contents` (modules) + `purge_exclusions` (things potentially generated by a post_run hook like `.generated_types` for environments). This is from the original Puppetfile specs.

mwaggett replied:
ahh okay, I think I understand. so then what if we said something like 'returns an array of directories that could potentially be purged'?

justinstoller replied:
Okay, updated locally to be:
"returns an array of paths that #purge! will operate within"

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+  end
+
+  describe 'evaluating a Puppetfile' do
+    def expect_wrapped_error(orig, pf_path, wrapped_error)
```
Would it make sense to call this param `error` instead of `orig`? And then maybe `wrapped_error` could be `wrapped_error_type`?

justinstoller replied:
yeah

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,297 @@
+      expect { subject.load! }.not_to raise_error
+    end
+
+    it 'creates a git module and applies the default branch sepcified in the Puppetfile' do
```
```suggestion
    it 'creates a git module and applies the default branch specified in the Puppetfile' do
```

mwaggett reviewed (commented):
I realize a lot of these comments are about things that were already in here, so feel free to ignore, but figured I'd mention everything 🙃

justinstoller committed:
(maint) Re-allow relative and absolute paths for moduledir & puppetfile

Previous to recent refactors an API user could provide either an
absolute or relative moduledir path, as well as an absolute
puppetfile_path or relative puppetfile_name, when resolving Puppetfiles.
The recent refactor required consumers to either pass absolute paths or
allow defaults rooted in the basedir, but disallowed passing relative
paths that would then be rooted in the basedir.

This commit returns that ability, as well as refactoring the internal
path joining and cleaning to be shared by all consumers of a file system
path.

This will change the the internal storage of basedir, moduledir, and
puppetfile to use a "cleaned" path stripped of extra slashes and dots.
However, this cleaned representation is what we would compare to
install_paths and display to users in error messages, so it should be a
change that does not affect any behavior.

justinstoller committed:
(maint) Clarify Puppetfile#add_module usage

justinstoller committed:
(maint) Minor cleanup of Puppetfile#add_module

justinstoller committed:
(maint) Minor renaming and comment clean up

This commit responds to several PR reviews for cleaning up local
variable names, usage of `!` in method names, clarifying comments and
minor test improvements.

justinstoller committed:
(maint) Rework default_branch_override passing for Puppetfile

Previously, the Puppetfile ModuleLoader would not extract and use the
default_branch_override from its options, which would cause the its
usage to fail.

We also would extract the wrong key for the default_override_branch in
R10K::Puppetfile, causing us always warn that the passed override and
the initialized one was wrong, even when they matched.

This patch also includes some minor fixes to how defaults are handled
and local variables are named.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,195 @@
+      # @param forge [String] The url (without protocol) to the Forge
+      # @param overrides [Hash] Configuration for loaded modules' behavior
+      # @param environment [R10K::Environment] The environment loading may be
+      #     taking place within
```
```suggestion
      # @param environment [R10K::Environment] When provided, the environment in which loading
      #     takes place
```
or something like that, to make clear that loading definitely (not may) happens within the environment, when that param is provided.

justinstoller replied:
accepted locally

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,195 @@
+
+        @managed_directories = determine_managed_directories(managed_content).freeze
+        @desired_contents = determine_desired_contents(managed_content).freeze
+        @purge_exclusions = determine_purge_exclusions(@managed_directories.clone).freeze
```
Why does this need to be cloned?

justinstoller replied:
The method that `determine_purge_exclusions` is based on used to mutate the argument. I did this, then updated it to no longer mutate the argument. I'll get rid of the clone.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -0,0 +1,195 @@
+          install_path = @moduledir
+        end
+
+        if @default_branch_override != nil
```
```suggestion
        if @default_branch_override
```

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +70,118 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
-    @default_branch_override = default_branch_override
+    if default_branch_override && (default_branch_override != @default_branch_override)
+      logger.warn("Mismatch between passed and initialized default branch overrides, preferring passed value.")
+      @loader.default_branch_override = default_branch_override
```
since GH doesn't seem to want to let me reply to my previous comment, Do we still want to set this if we can't read the Puppetfile and therefore never actually load?

Your reply referenced the conditional above, but I'm asking about the conditional below where we check to see if the Puppetfile is readable. I'm wondering if setting `@loader.default_branch_override = default_branch_override` should come after that check. Does that make sense? Apologies if I'm just misunderstanding your reply.

justinstoller replied:
No, that makes sense. Really my reply was "I don't care if it's not in the best spot because I want to deprecate and remove it when I can anyways". I updated the code earlier to try to be more clear in my intentions but I'll update the location in the meantime.

mwaggett commented on lib/r10k/puppetfile.rb:
```diff
@@ -67,142 +70,118 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
-    dirs = @managed_content.keys
-    dirs.delete(real_basedir)
-    dirs
+    @loader.managed_directories
```
Should this be `@loaded_content[:managed_directories]` instead of reaching into the `@loader`, like we do for `desired_contents` and `purge_exclusions` below? I assume both would work, but I figure we should be consistent unless there's a reason to prefer one over the other.

justinstoller replied:
good catch

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,330 @@
+                                                  '/test/basedir/',
+                                                  'env',
+                                                  { remote: 'git://foo/remote',
+                                                    ref: 'env' })
```
(nit) might be good to use a ref that's different from the environment name?

justinstoller replied:
The ref is usually the branch name that backs the environment, which is also the environment name. So while this is confusing, it is accurate, here's the code that creates Git environments (where `bn` represents the "branch name" object:
```rb
        envs << R10K::Environment::Git.new(bn.name,
                                           @basedir,
                                           bn.dirname,
                                           {remote: remote,
                                            ref: bn.name,
                                            puppetfile_name: puppetfile_name,
                                            overrides: @options[:overrides]})
```

mwaggett replied:
ahh, yeah, that makes sense, thanks

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,330 @@
+          expect(puppetfile.instance_variable_get(:@puppetfile)).to eq('/test/basedir/env/Puppetfile.global')
+        end
+      end
+
```
Could we move setting `subject` to down here, since the above tests don't use it?

justinstoller replied:
I don't know, do we have guidance around that?

I have a fairly ingrained habit of putting (and looking for) `before`s, `let`s, etc at the top of the describe/context block they're active for.

mwaggett replied:
I don't know of any guidance / style guide around that kind of thing. I'm fine leaving it if you prefer. My habit is to just look for things like that nearby the things that use it. 😅 Could just switch the order of the tests haha, but doesn't really matter.

Magisus replied:
I'd much rather `subject` be set at the top, that's definitely the way it's done everywhere I've seen.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,330 @@
+    subject { R10K::ModuleLoader::Puppetfile.new(basedir: basedir) }
+
+    it 'should transform Forge modules with a string arg to have a version key' do
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', subject.moduledir, hash_including(version: '1.2.3'), anything).and_call_original
```
apologies as it's been a while since I've had to write rspec tests, but why do we need this `allow`? We don't seem to be doing anything special that the original method doesn't do?

justinstoller replied:
I think the original author meant to use `expect` though once you put argument expectations on an `allow` I think it works essentially the same. I'll update this to `expect`.

mwaggett replied:
yeah, the `allow` vs. `expect` and how they exactly behave has always confused me, so using the one that semantically expresses the intention is my preference. 😁

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,330 @@
+      expect(subject.modules.collect(&:name)).to include('test_module')
+    end
+
+    it 'should not accept Forge modules with a version comparison' do
```
Is that just something we've never supported?

justinstoller replied:
I believe the Forge metadata and potentially librarian-puppet (an r10k precursor) did. But I think it has always been out of r10k's scope.

mwaggett replied:
gotcha, just making sure this wasn't new from this PR

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,330 @@
+
+    it 'should disable and not add modules that conflict with the environment' do
+      env = instance_double('R10K::Environment::Base')
+      mod = instance_double('R10K::Module::Base', name: 'conflict', origin: :puppetfile, 'origin=': nil)
```
What does this `'origin=': nil` bit do?

justinstoller replied:
each key becomes a method that takes any args and returns the value given. This defines a stub setter for `origin`.

mwaggett replied:
ahh, okay, so since we're stubbing this, `R10K::Module::Base` doesn't have its normal methods defined, and we need the `origin` setting for `add_module`. Is that an accurate understanding of why this is needed?

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -0,0 +1,330 @@
+    it 'should disable and not add modules that conflict with the environment' do
+      env = instance_double('R10K::Environment::Base')
+      mod = instance_double('R10K::Module::Base', name: 'conflict', origin: :puppetfile, 'origin=': nil)
+      loader = described_class.new(basedir: basedir, environment: env)
```
Any reason you use `described_class` here instead of directly instantiating `R10K::ModuleLoader::Puppetfile` like you do elsewhere?

justinstoller replied:
Nope! I'll normalize that.

justinstoller committed:
(maint) Respond to PR feedback

Magisus reviewed (approved):
I think I'm good on this.
